docs/interfaces.md

```diff
@@ -0,0 +1,568 @@
+
+## State:
+An ECC instance keeps a local reference of the FPC chaincode enclave id.
+
+```go
+type EnclaveChaincode struct {
+        erccStub ercc.EnclaveRegistryStub
+        tlccStub tlcc.TLCCStub
+        enclave  enclave.Stub
+        verifier crypto.Verifier
+        ev       attestation.EvidenceService // abstracts evidence creation for EPID/DCAP
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+
+# ECC_Enclave
+
+## Interface:
+The ECC_Enclave interface specifies interface of an FPC chaincode enclave. 
+
+Note: `sealed credentials` is renamed to `sealed state`
+
+```c++
+// initializes an enclave
+public int ecall_init(
+        [in, size=cc_params_len] uint8_t *cc_params, uint32_t cc_params_len,
+        [out] uint8_t *credentials);     // TBD is this a fixed length type?
+
+// invoke a FPC chaincode
+public int ecall_chaincode_invoke(
+        [in, size=proposal_len]  uint8_t *proposal, uint32_t proposal_len,
+        [out, size=response_len_in] uint8_t *response, uint32_t response_len_in,
+        [out] uint32_t *response_len_out,
+        [out] sgx_ec256_signature_t *signature,
+        [user_check] void *u_shim_ctx);
+
+// key generation and distribution
+// TODO interface below needs more brain cycles
+public int ecall_generate_cc_keys(
+        // TBD: Why do we need cc_params?
+        [in, size=cc_params_len] uint8_t *cc_params, uint32_t cc_params_len, 
+        [out] uint8_t *cckey_registration_msg);
+
+public int ecall_export_cc_keys(
+       [in] uint8_t *target_enclave_vk,
+       [out] uint8_t *export_msg);
+
+public int ecall_import_cc_keys(
+        // TBD: Why do we need cc_params?
+        [in, size=cc_params_len] uint8_t *cc_params, uint32_t cc_params_len, 
+        [in] uint8_t *export_msg,
+        [out] uint8_t *cckey_registration_msg);
+```
+
+```c++
+// Binding interface
+// note this is Post-MVP feature, but exists here fore completeness, TBD details
+public int ecall_get_CSR(
+        [in] args   
+        [out] csr);
+
+```
+
+## State:
+We also use the state, in particular, `internal_ecc_state` to control that `ecall_create_enclave` or `ecall_create_enclave_with_sealed_credentials` can be invoked once.
+
+TODO: change key types to whatever pdo crypto gives us.
+```c++
+// this is create on enclave creation and kept inside the enclave
+typedef struct internal_ecc_state_t {
+        cc_params_t cc_params              // inital chaincode parameter
+
+        // enclave-specific keys
+        sgx_ec256_public_t enclave_vk      // signature verification key
+        sgx_ec256_private_t enclave_sk     // signature key
+         
+        // chaincode-specific keys
+        sgx_aes_gcm_128bit_key_t sek,      // ledger state encryption key
+        sgx_ec256_public_t chaincode_ek,   // argument encryption key
+        sgx_ec256_private_t chaincode_dk,  // argument decryption key
+
+        // tlcc session
+        sgx_cmac_128bit_key_t session_key; // is set during tlcc-ecc binding
+};
+```
+
+# FPC_SHIM
```

Comment:
Per the lengthy conversation in, and the small code of, #405 , the `get_mspid` and `get_channel_id` APIs can also be implemented conveniently, thereby expanding the domain of supported chaincodes.

Reply:
added